src/environmentd/src/lib.rs

```diff
@@ -477,15 +477,24 @@ impl Listeners {
                     None,
                 )
                 .await?
+
+            // In read-only mode, we intentionally do not call `set_is_leader`,
+            // because we are by definition not the leader if we are in
+            // read-only mode.
         } else {
-            openable_adapter_storage
+            let adapter_storage = openable_adapter_storage
                 .open(
                     boot_ts,
                     &bootstrap_args,
                     config.controller.deploy_generation,
                     None,
                 )
-                .await?
+                .await?;
+
+            // Once we have successfully opened the adapter storage in
+            // read/write mode, we can announce we are the leader, as we've
+            // fenced out all other environments using the adapter storage.
+            deployment_state.set_is_leader();
```

Comment:
This looks much better to me. We still aren't quite ready to accept queries, we have to open the in-memory catalog among other things, but we have fenced out the previous version.

Reply:
Yup, exactly! And it's totally OK and correct IMO to report IsLeader even before we are able to accept queries, because we _are_ in fact the leader, and the old env can no longer accept queries (even though we can't either).